File: LeetCode Solutions/February 2022 LeetCoding Challenge/Swap Nodes in Pairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        if head == None:
            return None
        if head.next == None:
            return head
        if head.next.next == None:
            newHeader = head.next
            newHeader.next = head
            head.next = None
            return newHeader
        nextPairHeader = head.next.next
        newHeader = head.next
        newHeader.next = head
        head.next = self.swapPairs(nextPairHeader)
        return newHeader

File: LeetCode Solutions/Recursion I/Swap Nodes in Pairs/Swap Nodes in Pairs.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */


struct ListNode* swapPairs(struct ListNode* head){
    if (head == NULL){
        return NULL;
    }
    if (head->next == NULL){
        return head;
    }
    if (head->next->next == NULL){
        struct ListNode* newHeader = head->next;
        newHeader->next = head;
        head->next = NULL;
        return newHeader;
    }
    struct ListNode* nextPairHeader = head->next->next;
    struct ListNode* newHeader = head->next;
    newHeader->next = head;
    head->next = swapPairs(nextPairHeader);
    return newHeader;
}